### ISSUE TITLE
Fix explanation of evaluation order

### ISSUE THREAD

**Issue by florian-pe** [perl6/doc#4058]
The issue concers this page :
[https://docs.raku.org/language/operators](https://docs.raku.org/language/operators)

Here is pretty much the first sentence
```
The precedence and associativity of Raku operators determine the order of evaluation of operands in expressions.
```

Which is false, as demonstrated by this example : 
```
$ raku -e 'print(1) ** print(2) ** print(3); say ""'
WARNINGS for -e:
Useless use of "**" in expression "print(1) ** print(2) **" in sink context (line 1)
123
```
I made a github issure very recently about it, as I confounded operators precedence and associativity VS order of evaluation of operands. If there are no side effects, then it doesn't change anything and you wouldn't even know that the order is different, as a language user, but if there are side effects, then it becomes apparent that there is a distinction.

[https://stackoverflow.com/questions/71941798/why-does-print-isnt-affected-by-operator-precedence/71942002](https://stackoverflow.com/questions/71941798/why-does-print-isnt-affected-by-operator-precedence/71942002)

**Comment by coke** [comment]
The linked stackoverflow question is for perl, not raku.

**Comment by florian-pe** [comment]
@coke Yes, the stackoverflow question is about Perl.
But what's your point ? My raku example clearly shows that what the documentation is wrong. The stackoverflow question is about the EXACT SAME subject.

**Comment by florian-pe** [comment]
And I should add the problem is not that the documentation is wrong abour a random issue.
The problem is that it is wrong about the evaluation model of Raku. It is wrong about something so ridiculously fundamental that it's almost annoying to read when we know it is wrong.

**Comment by Altai-man** [comment]
B-but it should be (evaluating as stated in the docs), shouldn't it? I believe we want to ask why it works this way someone who knows the guts better to explain.

**Comment by florian-pe** [comment]
@Altai-man You're absolutely right, someone who knows the internals and/or the specifications should weight in to explain to us the rationale behind this.

I'm guessing that the raku implementation should behave as stated in the specification (at first RFCs + Apolcalypse, Exegesis, Synopsis, but now the "spec test" I believe). The docs aren't meant to be a specification I think, which is kind of a shame.

The Docs are really more a tutorial than a reference. And a lot of times the features are not described exhaustively.

I really think this is one of think that is unfriendly to beginners. Because if you need ed specifics about the semantics of a given feature, where else are you supposed to get your answer if it is not in the Docs ?

The Docs probably started with good intentions. They are meant for beginners, but not for intermediate/advanced users.

The Regex and Grammars sections are really incomplete for example. A lot of what is exposed in the Apocalypses/Synopsis/Exegesis are simply left out. Sometimes the specification changed, sometime features aren't yet implemented, sometimes features are implemented but not documented. We just don't know.

It's also kind of a shame considering that regexes are Perl's 🧬.

**Comment by lizmat** [comment]
```raku
    my Mu $methodcall       := nqp::hash('prec', 'y=');
    my Mu $autoincrement    := nqp::hash('prec', 'x=');
    my Mu $exponentiation   := nqp::hash('prec', 'w=', 'assoc', 'right');
...
trait_mod:<is>(&infix:<**>, :prec($exponentiation));
```
That makes exponentiation the tightest infix operator, right associative.

It may be late, but I think Raku is doing the right thing in this respect.  Am I missing something?

**Comment by thundergnat** [comment]
Raku is doing the right thing. (In this case at least.) The return value of a successful print statement is 1, (well... True, but 1 in numeric context,) so it is evaluating `1 ** 1** 1`, then not doing anything with it; hence the warning. 

Consider what happens if we override the print routine with something that returns its parameter (and if we use numbers that will generate a value other than 1):

```
sub print { $^a }                      # override print to return its supplied parameter
print(2) ** print(3) ** print(4);      # warns "useless use"
say print(2) ** print(3) ** print(4);  # as expected
say 2 ** 3 ** 4;                       # for comparison
```
output:

    Useless use of "**" in expression "print(2) ** print(3) **" in sink context (line 2)
    2417851639229258349412352
    2417851639229258349412352

Looks to me like it is doing the right thing.  I suspect the OP may be confused about what print() returns.

**Comment by 2colours** [comment]
Oh yes... I can recall a similar case [with the hyper operator](https://github.com/Raku/doc/commit/51eb039c8b98034dd8dce8ac8543b33719ed4c66).

**Comment by 2colours** [comment]
The quote, still present right at the beginning of the operators page:

> The precedence and associativity of Raku operators determine the order of evaluation of operands in expressions.

I tend to think that if we look deeper into this, this statement is false and perhaps a bit worse: conceptually wrong. Precedence and associativity is about the syntax tree - about what the operands of certain operations are to begin with. That's why "order of evaluation of operands" is a meaningful question for single operations (usually it doesn't matter but `a && b` for example is strictly left-then-right), even though "precedence" and "associativity" don't exist/make no sense in this context.

To give you an example: for `1 + 2 - 2 * 3`, by precedence, `2 * 3` is clearly (by the words of the current documentation) the "first sub-expression to evaluate", but I doubt anybody would argue that for this eventual subtraction, `2 * 3 == 6` **must** be evaluated before `1 + 2 == 3`. The important thing is just that the expression we have (now in Lisp notation) is `(- (+ 1 2) (* 2 3))`.

I don't know if "evaluation order" as such we already have at all as a topic, or if we need it at all. If we need it, I think it's going to be a new topic, whether within operators or outside, but it is clearly not tied to precedence and associativity. That sentence should be fixed in a way that it doesn't refer to order of evaluation at all.

**Comment by arkiuat** [comment]
> For "evalutation order (in Raku)", I'm not sure if the docs write about it

You just quoted the section where the docs write about it, so I'm not sure what you're meaning to say here. Evaluation order is discussed in the very first section of [Language/operators](https://docs.raku.org/language/operators).

The first two paragraphs of the [Operator precedence](https://docs.raku.org/language/operators#Operator_precedence) section explain how evaluation order is deduced from precedence and associativity.  The entries for each operator on that whole page are listed in precedence order for this reason, so the concept is fundamental to the entire structure and organization of Language/operators.rakudoc.

The Lisp example you give shows why Lisp doesn't need precedence rules. Precedence rules are only for people who'd like to write expressions without parentheses. Are you arguing that 1 + 2 - 2 * 3 should evaluate to something other than -3?

**Comment by 2colours** [comment]
Well, Raiph had a very exact proposal (one that I'm not 100% satisfied with but a clear proposal regardless), and what I originally had in mind was that my arguments about why precedence and associativity aren't about evaluation order at all would be understood, I would obtain a green light for making changes in that spirit, and then I would think about the exact phrasing. In my mind - without adding anything about _actual_ evaluation order which might deserve a new issue indeed - that would also be a couple of lines change that doesn't affect the overall structure of the document. Something along the lines "precedence determines _which operands belong to which operations_" or "(...) _what sub-expressions an expression contains_".
